### PR TITLE
Parallelize proposal transactions signature verification

### PR DESCRIPTION
**Pull request by tsachiherman**
## Summary

Our existing code was verifying the transactions in parallel with the insertion to the evaluator, but that won't fully utilize the host computational abilities. To push it to the end, we need to break the transaction verification into subsets and parallelize the verification.

This is the first PR, which would handle the verification parallelization. The subsequent PR would ensure that we don't revalidate the same transaction coming from two different proposals, regardless if the transaction appear in the transaction pool or not.

## Test Plan

Added unit test to verify changed code correctness.

## Performance

Testing the performance gains between running `verify.TxnPool` vs using the new `verify.PaysetGroups`:
| Transaction Count | Before ( serialized ) | After ( parallelized ) |
|:--:|:--:|:--:|
|1 (amortized) |65,200 ns| 14,600 ns|
|5,000|326 ms|73 ms|
|10,000|656 ms|146 ms|
|50,000|3.260 s|730 ms|

Tests were conducted on a MacBook Pro 2.7 GHz Quad-Core Intel Core i7

### CODE REVIEW

File: data/transactions/verify/txn.go [hunk 1]
```diff
@@ -314,3 +314,113 @@ func LogicSig(txn *transactions.SignedTxn, ctx *Context) error {
...
+		return nil
+	}
+	spec := transactions.SpecialAddresses{
+		FeeSink:     blk.BlockHeader.FeeSink,
+		RewardsPool: blk.BlockHeader.RewardsPool,
+	}
+	var txnGroups [][]transactions.SignedTxn
+	// prepare up to 16 concurrent worksets.
+	worksets := make(chan struct{}, 16)
+	worksDoneCh := make(chan interface{}, 16)
```
Comment (algorandskiy) [inline data/transactions/verify/txn.go:341]: why 16?

File: data/transactions/verify/txn.go [hunk 2]
```diff
@@ -314,3 +314,113 @@ func LogicSig(txn *transactions.SignedTxn, ctx *Context) error {
...
+								return err
+							}
+							if signTxn.Txn.Src() == zeroAddress {
+								return errors.New("empty address")
+							}
+
+							if err := stxnVerifyCore(&signTxn, &ctxs[k]); err != nil {
+								return err
+							}
+						}
```
Comment (algorandskiy) [inline data/transactions/verify/txn.go:399]: Looks the same as `verify.Txn` if `if !proto.SupportRekeying` moves into the loop.

File: data/transactions/verify/txn.go [hunk 3]
```diff
@@ -314,3 +314,113 @@ func LogicSig(txn *transactions.SignedTxn, ctx *Context) error {
...
+					txnGroupIdx = i
+					break
+				}
+				if i == len(payset)-1 {
+					txnGroups = payset[txnGroupIdx:]
+					txnGroupIdx = len(payset)
+					break
+				}
+				txnCounter += len(payset[i])
+			}
```
Comment (algorandskiy) [inline data/transactions/verify/txn.go:364]: possibly needs to be a method

File: ledger/eval.go [hunk 4]
```diff
@@ -995,40 +981,36 @@ func eval(ctx context.Context, l ledgerForEvaluator, blk bookkeeping.Block, vali
...
 				return StateDelta{}, err
-			default:
 			}
-			return StateDelta{}, ctx.Err()
 		default:
 		}
 
-		if validate {
-			txvalidator.txgroups <- txgroup
-		}
```
Comment (algorandskiy) [inline ledger/eval.go:1031]: another parallelism idea: run NumCPU-1 goroutines that equieue tasks received via `txvalidator.txgroups` channel. It possibly will have much less code but the same perf/parallelism.

**Review by algorandskiy — COMMENTED** [review]
The code (especially `verify. PaysetGroups`) is quite hard to follow, please use existing `verify.Txn` function in the pool's task.

**Comment by algorandskiy** [comment]
I'd love to see detailed explanation in the commit message like "split entire list of txn groups to X blocks and post them to processing in parallel".

File: ledger/eval.go [hunk 5]
```diff
@@ -1043,10 +1025,15 @@ func eval(ctx context.Context, l ledgerForEvaluator, blk bookkeeping.Block, vali
 
 	// If validating, do final block checks that depend on our new state
 	if validate {
-		close(txvalidator.txgroups)
-		err, gotErr := <-txvalidator.done
-		if gotErr && err != nil {
-			return StateDelta{}, err
+		// wait for the validation to complete.
+		select {
+		case err, open := <-txvalidator.done:
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply ledger/eval.go:1030]: no, if you'll look at the code, you'll see that it closes the channel when it's done and there is no validation error.
You could ask then, why do we need `if err != nil {`, since we already know it would be a non-nil; but I don't really think it adding that extra `if` statement would hurt much here.

File: ledger/eval.go (hunk 4 above)
Comment (tsachiherman) [reply ledger/eval.go:1031]: The execution pool is already using the NumCPU; that's why we use it..

File: data/transactions/verify/txn.go (hunk 2 above)
Comment (tsachiherman) [reply data/transactions/verify/txn.go:399]: yes, but placing it outside the loop is more efficient as it's being executed once per transaction group rather then once per transaction. And thinking about that, I should take it one loop upward since it's *normally* not executed.

File: data/transactions/verify/txn.go (hunk 1 above)
Comment (tsachiherman) [reply data/transactions/verify/txn.go:341]: I'll add const to define these with some explanation.

File: data/transactions/verify/txn.go [hunk 6]
```diff
@@ -314,3 +314,113 @@ func LogicSig(txn *transactions.SignedTxn, ctx *Context) error {
...
+	}
+	var txnGroups [][]transactions.SignedTxn
+	// prepare up to 16 concurrent worksets.
+	worksets := make(chan struct{}, 16)
+	worksDoneCh := make(chan interface{}, 16)
+	zeroAddress := basics.Address{}
+	processing := 0
+	tasksCtx, cancelTasksCtx := context.WithCancel(ctx)
+	defer cancelTasksCtx()
+	for processing >= 0 {
```
Comment (brianolson) [inline data/transactions/verify/txn.go:360]: It's clever that producer and consumer are both in this loop, but I think it would have been clearer split out. Probably spawn a go routine for the producer and then run the consumer loop in this function below that.

File: data/transactions/verify/txn.go (hunk 3 above)
Comment (tsachiherman) [reply data/transactions/verify/txn.go:364]: I think it would make it quite confusing to have it as a separate function, given that it would need to be something like:
```
func getNextWorkset(payset [][]transactions.SignedTxn, txnGroupIdx *int) (txnGroups [][]transactions.SignedTxn) {
                        txnCounter := 0 // how many transaction we already included in the current workset.
			// scan starting from the current position until we filled up the workset.
			for i := *txnGroupIdx; i < len(payset); i++ {
				if txnCounter+len(payset[i]) > txnPerWorksetThreshold {
					if i == *txnGroupIdx {
						i++
					}
					txnGroups = payset[txnGroupIdx:i]
					*txnGroupIdx = i
					return
				}
				if i == len(payset)-1 {
					txnGroups = payset[*txnGroupIdx:]
					*txnGroupIdx = len(payset)
					return
				}
				txnCounter += len(payset[i])
			}
                        return nil
} 
```

File: ledger/eval.go [hunk 7]
```diff
@@ -923,58 +924,43 @@ func (eval *BlockEvaluator) GenerateBlock() (*ValidatedBlock, error) {
...
+	defer close(validator.done)
+	verifyParams := verify.Params{
+		CurrSpecAddrs: transactions.SpecialAddresses{
+			FeeSink:     validator.block.BlockHeader.FeeSink,
+			RewardsPool: validator.block.BlockHeader.RewardsPool,
+		},
+		CurrProto: validator.block.BlockHeader.CurrentProtocol,
+	}
+	var unverifiedTxnGroups [][]transactions.SignedTxn
+	unverifiedTxnGroups = make([][]transactions.SignedTxn, len(validator.txgroups))
```
Comment (brianolson) [inline ledger/eval.go:944]: this was probably supposed to be the 3 arg form make(type, len, cap) with len=0

File: ledger/eval.go [hunk 8]
```diff
@@ -1043,10 +1025,15 @@ func eval(ctx context.Context, l ledgerForEvaluator, blk bookkeeping.Block, vali
 
 	// If validating, do final block checks that depend on our new state
 	if validate {
-		close(txvalidator.txgroups)
-		err, gotErr := <-txvalidator.done
-		if gotErr && err != nil {
-			return StateDelta{}, err
+		// wait for the validation to complete.
+		select {
```
Comment (brianolson) [inline ledger/eval.go:1029]: don't need a one option select{}, just do the receive statement?

File: data/transactions/verify/txn.go (hunk 6 above)
Comment (tsachiherman) [reply data/transactions/verify/txn.go:360]: I don't think that making the producer a separate go-routine makes sense here. The time we're spending on the producer is by order of several magnitudes smaller than the consumer. i.e. the loop is waiting in the select statement almost fo the entire time.

File: ledger/eval.go (hunk 7 above)
Comment (tsachiherman) [reply ledger/eval.go:944]: good catch.

File: ledger/eval.go (hunk 8 above)
Comment (tsachiherman) [reply ledger/eval.go:1029]: hmm.. no, it's a bug:
it should also test for `ctx.Done()` since the context might be canceled. Will be fixed shortly.

File: data/transactions/verify/txn.go [hunk 9]
```diff
@@ -314,3 +329,116 @@ func LogicSig(txn *transactions.SignedTxn, ctx *Context) error {
...
+		case <-tasksCtx.Done():
+			return tasksCtx.Err()
+		case worksets <- struct{}{}:
+			if len(txnGroups) > 0 {
+				err := verificationPool.EnqueueBacklog(ctx, func(arg interface{}) interface{} {
+					// check if we've canceled the request while this was in the queue.
+					if tasksCtx.Err() != nil {
+						return tasksCtx.Err()
+					}
+					txnGroups := arg.([][]transactions.SignedTxn)
```
Comment (algorandskiy) [inline data/transactions/verify/txn.go:393]: I still think `verify.Txn` should be called here inside loops. most likely the benchmark will not show any noticeable difference but the code would be clearer

File: data/transactions/verify/txn.go (hunk 3 above)
Comment (algorandskiy) [reply data/transactions/verify/txn.go]: or
```
type workset struct {
  payset [][]transactions.SignedTxn
  idx int
}

func (w *workset) next() (txnGroups [][]transactions.SignedTxn) {
...
}

func (w *workset) completed() bool {
  return w.idx >= len(w.payset)
}
```

File: data/transactions/verify/txn.go (hunk 2 above)
Comment (algorandskiy) [reply data/transactions/verify/txn.go]: I suggest to measure. Benefits: better code, no duplicate code.

File: data/transactions/verify/txn.go (hunk 3 above)
Comment (tsachiherman) [reply data/transactions/verify/txn.go]: that looks much better; I'll add that.

File: data/transactions/verify/txn.go [hunk 10]
```diff
@@ -314,3 +335,124 @@ func LogicSig(txn *transactions.SignedTxn, ctx *Context) error {
...
+			return tasksCtx.Err()
+		case worksets <- struct{}{}:
+			if len(nextWorkset) > 0 {
+				err := verificationPool.EnqueueBacklog(ctx, func(arg interface{}) interface{} {
+					// check if we've canceled the request while this was in the queue.
+					if tasksCtx.Err() != nil {
+						return tasksCtx.Err()
+					}
+					txnGroups := arg.([][]transactions.SignedTxn)
+					if !proto.SupportRekeying {
```
Comment (algorandskiy) [inline data/transactions/verify/txn.go:379]: Current
```
BenchmarkPaysetGroups-8   	   78248	     16482 ns/op
BenchmarkPaysetGroups-8   	   75268	     16444 ns/op

BenchmarkTxnPool-8   	   18794	     63580 ns/op
BenchmarkTxnPool-8   	   18135	     64093 ns/op
```

With vanila `Txn(...)` call and the only one two loops instead of 4:
```
					txnGroups := arg.([][]transactions.SignedTxn)
					for _, signTxnsGrp := range txnGroups {
						ctxs := PrepareContexts(signTxnsGrp, blk.BlockHeader)
						for k, signTxn := range signTxnsGrp {
							if err := Txn(&signTxn, ctxs[k]); err != nil {
								return err
							}
						}
					}
```
```
BenchmarkPaysetGroups-8   	   80364	     15178 ns/op
BenchmarkPaysetGroups-8   	   82190	     14846 ns/op

BenchmarkTxnPool-8   	   18781	     66758 ns/op
BenchmarkTxnPool-8   	   18808	     63886 ns/op
```

Comment (algorandskiy) [reply data/transactions/verify/txn.go:379]: I guess just calling `Txn` is desirable since it is not slower and gives twice less code and no code duplication.

**Review by brianolson — DISMISSED** [review]
Makes my benchmarks in #1743 faster. Seems good.

**Review by algorandskiy — APPROVED** [review]
